### PR TITLE
discv5: expand topic index description

### LINKED ISSUE #110: discv5: need more info on global topic queue limit
The topic discovery mechanism has a global limit on topic advertisements:

> There is also a global limit on the number of ads regardless of the topic queue which contains them. When the global limit is reached, the last entry of the least recently requested topic queue is removed.

There are three possible issues with this that I can see:

1.  An attack where malicious nodes can conspire to continuously register random, new topics, and request the ads placed in those topics to make them no 'least recently requested'
2. A similar attack but where the topic registrations are made concurrently, as if I understand correctly there is no prevention over issuing concurrent ad requests for 1000 different new topics (eg)
3. Legitimate advertisers placing ads on a server with the global limit at threshold may find their ads being selected for immediate removal.

### LINKED ISSUE #111: discv5: ticket hoarding attack
There is a really simple attack on the topic queue logic the way it is defined right now:
The wait period formula is supposed to keep ads in the queue for certain amount of time. 
Since the new wait period depends only on the time of the last successful registration,
an attacker controlling many nodes may request lots of tickets (up to the queue limit) on the same queue concurrently. Those tickets will expire around the same time and can be used to replace all queue content. When this happens, the wait period will jump up to a very high value,
i.e. not only will the queue be full with attacker-controlled nodes, the nodes will also stay in the queue for a long time.

### LINKED ISSUE #112: discv5: topic radius consensus problem
Both searching and advertising rely on the 'radius estimation' to figure out which nodes
are supposed to keep registrations for a certain topic. To estimate the radius from
received tickets, the node checks how closely the time in received ticket matches the
`target-ad-lifetime` value, a protocol constant. A big issue with this approach is that all nodes must agree on a certain constant, otherwise the algorithm doesn't converge properly. Getting the radius right is important, and small implementation differences can mess up the system.

I think it'd be better to estimate the radius based on the rate of successful registrations within
a region of address space. That is, instead of trying to converge on the optimum waiting time,
maintain a table of address buckets with a running average of the time it took to actually get
registered in that region.

### PR DESCRIPTION
**Pull request by fjl**
This change adds a lot more text about the topic subsystem and
changes the waiting time algorithm to ensure registration is resilient
against the recently discovered ticket hoarding attack. There is still
no simple and nice solution for radius estimation.

Closes #110 
Closes #111 
Closes #112

### CODE REVIEW

File: discv5/discv5-theory.md [hunk 1]
```diff
@@ -92,141 +92,199 @@ from the `k` closest nodes it has seen.
...
+
+- When the table is full, the waiting time is assigned based on the lifetime of the oldest
+  ad across the whole table, i.e. the registrant must wait for a table slot to become
+  available.
+- When the topic queue is full, the waiting time depends on the lifetime of the oldest ad
+  in the queue. The assigned time is `target-ad-lifetime - oldest-ad-lifetime` in this
+  case.
+- Otherwise the ad may be placed immediately.
+
+Tickets are opaque objects, not unlike HTTP session cookies. While details of encoding and
```
Comment (holiman) [inline discv5/discv5-theory.md]: Http session cookies are, by definition, lookup keys to a server-side session. 
You maybe mean some other type of cookie, where the state itself is encoded into the cookie? That's what I understand from context further down

File: discv5/discv5-theory.md [hunk 2]
```diff
@@ -92,141 +92,199 @@ from the `k` closest nodes it has seen.
...
-Node `A` now waits for the duration of the wait period. When the wait is over, `A` sends a
-registration request including the ticket. `C` does not need to remember its issued
-tickets, just the serial number of the latest ticket accepted from `A` (after which it
-will not accept any tickets issued earlier).
+    A <- C  TICKET [ticket, wait-time]
 
-    A -> C  REGTOPIC
+Node `A` now waits for the duration of the waiting time. When the wait is over, `A` sends
+another registration request including the ticket. `C` does not need to remember its
+issued tickets since the ticket is authenticated and contains enough information for `C`
```
Comment (holiman) [inline discv5/discv5-theory.md]: Here you claim that `C` does not need to remember tickets, which then assumes that there is no server side 'session'

Comment (fjl) [reply discv5/discv5-theory.md]: Yeah, maybe the HTTP session analogy isn't the best. I'll replace it with something else.

Comment (fjl) [reply discv5/discv5-theory.md]: Tickets are more like JWTs

**Pull request merged**